### DISCUSSION TITLE
Allow app to sync data across devices

### DISCUSSION THREAD

**Discussion by iSoron** [iSoron/uhabits#104] [Feature Requests]
It would be nice if Loop could synchronize your habit data across multiple devices. This feature has been one of the most frequently requested so far. It also opens many interesting possibilites, such as syncing data with a future desktop or web client.

Since this feature raises many privacy and security concerns, here is my non-comprehensive list of requirements:
1. It should be completely opt-in, and the app should still be fully functional with this feature disabled.
2. If there is software running on a server to make this feature work, this software should be fullly open-source and self-hostable.
3. It goes without saying, but the whole thing should just work. In particular, it must gracefully handle the case where multiple devices make changes while offline, then go online. The solution must be able to merge the offline changes, not simply discard all the changes made all clients, except one.

Because of requirement 2, most commonly used solutions such as Parse, Firebase or Play Games Services are out of the picture. Requirement 3 probably eliminates file-based sync solutions, such as Syncthing, Google Drive or Dropbox.

**Comment by iSoron** [comment]
<s>I am currently experimenting with custom back-end solution, implemented in node.js, on the branch `feature/sync`, but it's not usable yet.  I will post here once it works somewhat properly.</s>

Other ideas are very welcome, by the way.

**Comment by NiklasMM** [comment]
Do you want to host that backend yourself and make it available to all users? Or would that be a self-hosted solution, which everyone could set-up for themselves?

**Comment by iSoron** [comment]
Hi @NiklasMM, I will probably host an instance myself and make it available for all users. I might start accepting donations to keep the server running. In any case, you will also be able to run a private instance of the server yourself.

**Comment by quark6789** [comment]
Would it be possible to use git for this? Have each device make its own little repo with the data in it, and then either have each sync with a server, or possibly sync directly with each other (decentralized).

For merging, only conflicts for the same habit and day would be an issue.  I think choosing the most recent change would be a fine solution.

**Reply by colonelpanic8** [reply]
THIS. git is quite good at managing conflicts as well.

**Comment by iSoron** [comment]
@PunctualEmoticon If you are suggesting to create a local git repository on each phone with the database file, then I think we would run into the same problems as using Syncthing or Dropbox; namely, there is not enough granularity for git to merge things automatically. Only if we split the database into multiple files, would git (or Syncthing/Dropbox) be able to help us, but this doesn't strike me as a very good solution.

**Comment by quark6789** [comment]
Yes, that's what I was suggesting.
But unlike those file syncing services, git isn't limited to just checking files as a whole.  It looks at the contents of the files (if they're ultimately text files), and can resolve changes automatically, unless the same line has been modified on both branches since the branches last split.
Source: http://www.drdobbs.com/tools/three-way-merging-a-look-under-the-hood/240164902

**Comment by xdmx** [comment]
You may consider the ability to sync it with a nextcloud/owncloud server, that'd be nice

**Comment by anteloio** [comment]
I believe you don't want to have any server structure on your side. So, as you suggested above, maybe we could try to use Dropbox. 

Take a look at 1password for example. It works like a charm across my devices (mobile and laptop) and it's always in sync by using Dropbox as "database". Do you believe that might be feasible?

**Reply by LeonardoGentile** [reply]
Another nice thing about 1password is the ability to sync over wifi. 
I don't use dropbox to sync, but the app has a self-hosted sync server embedded in the desktop app. When connected to the same wifi I can sync from desktop to mobile and vice-versa

**Comment by kyriakosbrastianos** [comment]
What about webdav? I would like to share and use data with apps like nextcloud tasks & calendar and qownnotes: https://github.com/pbek/QOwnNotes 
QOwnNotes is great on desktop and syncs great with nextcloud tasks so u get web interface too and loop habits is amazing on android. A collaboration will be great with these services/apps.

**Reply by Higurashi-kagome** [reply]
> What about webdav? I would like to share and use data with apps like nextcloud tasks & calendar and qownnotes: [pbek/QOwnNotes](https://github.com/pbek/QOwnNotes) QOwnNotes is great on desktop and syncs great with nextcloud tasks so u get web interface too and loop habits is amazing on android. A collaboration will be great with these services/apps.

I also support implementing WebDAV. Doing so would enable users to sync their data to any platform they want, provided it supports WebDAV.

**Comment by omarbenhamid** [comment]
What about using cloudant, it is à cloud service based on opensource couchdb technology ? 
We could use https://github.com/cloudant/sync-android library. It should transparently fit the 3 requirements.

**Comment by MaluNoPeleke** [comment]
An optional Play Games Services support would be nice, even if it doesn't meet all requirements.

**Comment by recheej** [comment]
What about a web based solution? Send data back and forth between the web server. The web server stores the information in a database.

**Comment by mr-flibble** [comment]
Hello, just an idea: 
Since now application uses sqlite, would it not be too time-consuming to allow a user to attach an application to their own mysql postgresql server? The application would be in the online only mode but would be available from multiple devices.

**Comment by jharder01** [comment]
I'm in favor of self hosted options as they are in-line with the spirit of this project. Granted, a shared, public sync server may help instead awareness of the app. 

Additionally, I would love to be able to sync my Android Loop data locally to a personal Grafana/InfluxDB server for data logging and visualization.

**Comment by namdnguyen** [comment]
Where is the application sqlite database, not backup files, stored on Android? I'm having trouble finding it.

**Comment by eppeters** [comment]
I have the same question as namdnguyen

**Comment by jbrinker** [comment]
What about sync'ing up with third party services such as Gmail's Tasks or Drive. This way you would not have to worry about hosting back-end service endpoints for data sync'ing and storage.  You would also not have to burden yourself with the cost of hosting.

**Comment by trymeouteh** [comment]
This would be cool. The best way to do this is to not use a server but link each device together in a P2P way. Kinda like SyncThing.

Also it will need to encrypt the data on transit to prevent any tampering or having it being intercepted in transit.

**Comment by bendespain** [comment]
Why not use Google drive like this says: https://stackoverflow.com/questions/30233198/how-to-save-app-data-on-google-drive
with this direction: https://developers.google.com/drive/android/appfolder
and this: https://developers.google.com/drive/android/files#writing_files

If this creates conflicts then just make the app (if cloud-enabled is setup/chosen) wait until the drive connection is up, before it is allowed to write. Then Google drive is responsible for any conflicts, right?

**Comment by bendespain** [comment]
Or maybe just syncing it to Google Calendar?

**Comment by LoopTrooper** [comment]
This feature would be awesome!!

**Comment by trymeouteh** [comment]
Google Drive would not be good for privacy. It would be best to have the two devices talk to eachother.

If this feature is added then people can have their habits synced between their tablets and phones like many other apps.

**Comment by CryptoStar2000** [comment]
Awesome! I would appreciate this!!

**Comment by TheHolyLoli** [comment]
Any update on this @iSoron ?
also, i cant find `features/sync` branch.is it deleted?

**Comment by LoopTrooper** [comment]
Still have huge interest in this feature :-)!!

**Comment by CryptoStar2000** [comment]
Is the app still in development? Sync would be so great. Or at least the ability to save and restore in cloud.

**Comment by trymeouteh** [comment]
I have a suggestion on how to make the sync feature easy to implement.

Have it so the habits is saved as a file on your android device. Every time you add/remove a checkbox, add or remove a habit it will update the file and save it. All that is needed now is for the user to setup Syncthing or their cloud storage app to sync the data between their devices. This method will not require adding a syncing feature to Habits and allows the users to control their data by letting them choose what services to sync their data with. A good app for cloud storage syncing is FolderSync and there are other apps like FolderSync out there.

By going down this method you will also be for filling the automatic backup feature request too.

**Comment by hiqua** [comment]
@iSoron can you expand on what should be synchronized, to understand better why Syncthing (in particular) is not an option? I have this in mind:

* name of the habit
* frequency
* reminder
* archived status
* days on which it has been completed
* color

Let me know if I'm forgetting something.

Now, for an habit, you could have a corresponding folder (folder name is a UUID which does not change), containing two subfolders
* `journal`: contains an empty file like `2019_07_11` for every day the habit has been completed.
* `settings`: contains `name`, `frequency`, `reminder` and `color` with a timestamp appended to these names, containing the corresponding parameters. The app would only take the last one into account (and possibly remove the old ones, or keep them as a backup to revert changes).

The entries in `journal` would not create real conflicts (e.g., if you tick on two devices then untick on one device then sync, then (with syncthing) the habit would end up being ticked). You could also use the same kind of timestamp system if need be, but I don't think it's necessary. For `settings`, no file conflict either, but the last option will be kept.

If you create two habits with the same name on different devices, they won't be merged, they will be separate habits, which I think is fine.

The main problem is that the current format is sqlite (AFAIK), so that would be a change there. I assume there's some performance penalty. To me, having a simple structure for which I don't need yet another tool to synchronize my data between devices would be a big win. I also don't think the performance penalty is that huge.

I'd be happy to learn about shortcomings of this approach. Again, the benefits is being able to synchronize data just by:
* changing the current data format
* using a synchronization tools like the ones you've mentioned in your first post

**Comment by iSoron** [comment]
@hiqua That's an interesting solution. Please feel free to code a prototype, but I'm guessing that you will run into several issues:
1. This requires users to install and maintain an external piece of software to make the feature work. I expect most users will find this too complicated, and simply never use the feature. I'm fine with features for power users (e.g. Tasker support) but I feel this one should be accessible to everyone.
2. This implementation would require Loop to store the database in a public folder. Besides the security and privacy implications of this decision, it should also be noted that Android in general is [moving away from such practice](https://developer.android.com/preview/privacy/scoped-storage). It's not clear if this solution would still work after Android Q.
3. Loop would need to monitor the filesystem for external changes and update the JVM objects accordingly. If the filesystem and JVM objects ever get out of sync, the app could crash trying to write into non-existing folders, or something similar. This also opens up the door to all kinds of race conditions. What happens if the folder is deleted externally between the time we check its existence and the time we read its contents?
4. This solution is incompatible with future web clients.

**Comment by hiqua** [comment]
Thans for taking the time to review my comment. I'll try to clarify possible solutions point by point:

1. This does not have to exclude any other in-app synchronization feature. My main point was that, by changing the data format (from sqlite), you could make it friendlier to dumb file synchronization apps, which would partially resolve this issue for some users who already use such apps. An in-app synchronization feature could still be implemented (almost) independently.
An example of an app that gives people the choice as to how synchronization is done is simpletask: https://github.com/mpcjanssen/simpletask-android. It might look complicated with their multiple build systems, but this doesn't have to be this way. There might be other advantages of sqlite that I'm not aware of though.

2. I think it's fair to keep the files in Loop's private folder, and leave it to synchronization apps and their users as to how they want to synchronize this data. Also note that, as of now, apps like WhatsApp (which arguably has more private content than Loop) still store a lot of their data in the common (public) storage. Once (if) public folders are deprecated, I'm confident workarounds will come up to still have some kind of public folders, or at least to share these folders via synchronization apps.

3. I have been trying to find such conflicts, but I haven't found real ones yet. To "delete" a habit a `deleted` file can be added into the folder, instead of actually deleting any file or folder in the file system. If somebody manually deletes a folder, then all bets are off of course (but it's the same if somebody opens the sqlite database manually and injects some weird data in it).
Files and folders don't have to be monitored if the user is fine only having eventual consistency (e.g., by rescanning the files every 24h). To always have consistent data across devices, there's indeed some monitoring required, be it of files with or with a connection to a server with the solution you currently have in mind. I'm not sure if one is clearly better than the other (I'm assuming the inotify API is available on Android, but I don't really know).

4. I admit I am not aware of the requirements of future web clients.

**Comment by MoSattler** [comment]
Hi @hiqua, did you make any progress on this?

**Comment by hiqua** [comment]
@MoSattler no.

**Comment by mdaudali** [comment]
Hey! Just thought I'd add my 2 pence here - from a theoretical point of view, Conflict Free Replicated Data types would be the way to go (https://crdt.tech/ for some background reading). These data structures allow each node / device to operate over the data with defined operations (e.g, marking the task as completed), and have defined serialisation rules that ensure multiple nodes can merge their operations together, resulting in a consistent end view. 

From a practical perspective, this can be implemented in a couple of different ways. My undergrad dissertation (feel free to ask for a copy) used an OCaml library called Irmin that actually interfaces with Git, but manages the underlying merge semantics for you. An example structure for this project could be a simple JSON structure for each habit, with intervals marking the dates that have been marked complete. The 3 way merge would simply merge these intervals, always favouring completed over uncomplete. Devices would simply `git push` to a server, letting the server handle the merge.

This would satisfy your second and third requirements - For example, it would likely be under 100 lines of OCaml and easily made open source. By definition, it would handle multiple devices syncing at any point in time, even with potentially conflicting changes. It'd also make it really easy to backup, since backing up is simply git clone.

If I've not explained myself well, do feel free to ask for clarification! It's actually quite simple to implement, but may require a change in how the data is stored. I may try and give it a go over the next few months if I get time.

**Comment by tobiasgrosser** [comment]
> Hey! Just thought I'd add my 2 pence here - from a theoretical point of view, Conflict Free Replicated Data types would be the way to go (https://crdt.tech/ for some background reading). These data structures allow each node / device to operate over the data with defined operations (e.g, marking the task as completed), and have defined serialisation rules that ensure multiple nodes can merge their operations together, resulting in a consistent end view.
> 
> From a practical perspective, this can be implemented in a couple of different ways. My undergrad dissertation (feel free to ask for a copy) used an OCaml library called Irmin that actually interfaces with Git, but manages the underlying merge semantics for you. An example structure for this project could be a simple JSON structure for each habit, with intervals marking the dates that have been marked complete. The 3 way merge would simply merge these intervals, always favouring completed over uncomplete. Devices would simply `git push` to a server, letting the server handle the merge.
> 
> This would satisfy your second and third requirements - For example, it would likely be under 100 lines of OCaml and easily made open source. By definition, it would handle multiple devices syncing at any point in time, even with potentially conflicting changes. It'd also make it really easy to backup, since backing up is simply git clone.
> 
> If I've not explained myself well, do feel free to ask for clarification! It's actually quite simple to implement, but may require a change in how the data is stored. I may try and give it a go over the next few months if I get time.

This overall sounds quite interesting, but I would assume that OCAML cannot be really shipped easily to Android. I wonder if this could be done without the need for OCAML?

**Comment by mdaudali** [comment]
The OCaml would only be needed on the server side - the communication protocol is Git or even a REST API and so is language agnostic, so the app can use whatever is currently in use. 

OCaml also isn't necessary for the server side - I mention it as it's what I used for my project and there's already a well tested library (https://irmin.org/) for CRDTs. However, it can be implemented in any language - it'd just take a bit more work to implement.

**Comment by tobiasgrosser** [comment]
Nice, this sounds great for my perspective. I am not the project lead, but from experience, I feel that some proof of concept code typically helps to kick things off. I would be super interested to see this feature.

**Comment by iSoron** [comment]
@mdaudali @tobiasgrosser A proof of concept would be very welcome, but I hope this solution doesn't require one git repository for each active user, otherwise it wouldn't scale very well. I estimate that Loop is currently installed in close to a million devices.

**Comment by falcorr** [comment]
I'd be happy even with simply having an option to sync with Google Drive, Dropbox, whatever.

**Comment by TheHolyLoli** [comment]
nextcloud/owncloud/webdav support is also good.

**Reply by raffh** [reply]
Indeed, being able to auto-sync with Nextcloud would be amazing.

**Reply by oakbrad** [reply]
CalDAV / WebDAV sync seems like the most obvious and widely supported solution. Both calendar entries and tasks are obviously natively supported by iOS/Android, the syncing is already built into your phone and software, and the server backend is already easily and common self hosted if you don't trust cloud. 

Even just having Loop output a read-only ICS would be useful, but CalDAV sync as the backend would really be amazing as it would make Loop's information available in everyday productivity software.

**Comment by ramzan** [comment]
What about [Etebase](https://www.etebase.com/)? It's the backend used in [Etesync](https://www.etesync.com/).

**Reply by SkypLabs** [reply]
I was about to suggest this option.

I think Etebase ticks most of the boxes:

- Privacy-friendly (end-to-end encryption).
- Possibility to deploy your own server or to use someone else's without compromising your privacy thanks to E2E.
- Cross-device synchronisation with conflict handling and versioning.
- Offline support.
- Etebase back-ends can be mutualised to store and synchronise the data of different projects. No need to set up a server specifically for uhabits.
- Project widely audited.

**Comment by iSoron** [comment]
@ramzan Thank you for the link. That actually looks quite promising. We should look into it.

**Comment by iSoron** [comment]
A partial implementation of this feature has been merged into the `dev` branch and will probably be available in Loop 2.0, unless we find serious bugs or limitations in our implementation. (UPDATE: This feature still requires a bit of work, so we are delaying it to Loop 2.1)

If you would like to try out an early work-in-progress version of this feature and give us some feedback, it is also available in [v2.0.0-alpha](https://github.com/iSoron/uhabits/releases/). Please note that alpha releases may contain serious bugs that can cause you to lose your entire data. You should create a full database backup before trying it out.

**Reply by snpefk** [reply]
I tried to open link from sync option (e.g. `https://loophabits.org/sync/<some hash>`) and got 404 😢

**Reply by iSoron** [reply]
@snpefk Are you trying to open this link on a phone that has Loop 2.0.0-alpha installed? You should have been offered the option to "Open with Habits". [Feel free to open a bug report](https://github.com/iSoron/uhabits/issues/new/choose).

**Reply by snpefk** [reply]
Oh, so it's an app link? I thought that was a just link for download a sync archive. I wanted to use it in my scripts that load data to my Grafana backend.

**Reply by iSoron** [reply]
@snpefk Yes, it's an app link; but even if you were able to download the database from the sync server, it would be unreadable (it's an encrypted file, with the decryption key stored only in your phone).

**Reply by snpefk** [reply]
> but even if you were able to download the database from the sync server, it would be unreadable

Can we mitigate this? For example with chechbox (and warning) for disabling encryption. I want to sync my data with my PC, move csv file from time to time seems clumsy.

**Comment by lkraav** [comment]
> A partial implementation of this feature has been merged into the dev branch and will probably be available in Loop 2.0, unless we find serious bugs or limitations in our implementation.

Tyvm :pray: Is there a short summary of how it technically works, somewhere, or gotta dig through code?

**Reply by iSoron** [reply]
Not at this time, since it's very experimental and subject to change, but I am planning to document it when the protocol gets more stable. For now, please see the source code.

**Comment by olegivo** [comment]
I think should be a core functional with high level sync logic and "plugin" modules for sync with concrete data sources such as WebDav or Dropbox or custom backend and so on.

**Comment by spiraldancing** [comment]
Quick thoughts on this feature idea.  Anki / Ankiweb might offer a good model to base the design on.  Also, uHabit already allows for data backup/restore ... a simple(r) first-draft solution would be expanding that tool to handle a relatively seamless behind-the-scenes automatic backup-restore process through a Cloud storage location of the user's choice, be it Dropbox or a self-hosted Nextcloud or whatever.  If it isn't already, data backup file should/oughta be encrypted for this.

... my dos centavos.

**Comment by VA-MMXXII** [comment]
There appear to be a number of great synching options out there. I've used ​the open source Tasks.org app on Android (https://play.google.com/store/apps/details?id=org.tasks&hl=en_US&gl=US) and it offers a number of ways to synch. Maybe one or more of them would allow Loop users to use their phones or their tablets (or maybe even their desktop) to check off habits throughout the day?

**Reply by Improver2000** [reply]
Same here! Tasks sync works perfect! Implementation would be wonderful :-)

**Comment by fel1** [comment]
I've noticed that many of the Habits that I've added are more like to-do's/chores, which are repeatable. I like the app functionality and simplicity so much, that now many of my daily basic to-do's also live here. I.e. "Take out the trash", "Feed the dog" etc.

It would be more than amazing to be able to share some habits (tasks) with my wife so either of us could mark them as done on any of our devices. This would need some sort of indicator of who (which user) ticked off the habit. If some sort of "Users" function could be added then a file-based sync solution might work well, depending on the implementation. What do you think?

**Comment by recheej** [comment]
Hey @iSoron just checking the status of this one. Are you still looking for help with it?

**Comment by woodsea9** [comment]
I would love this to be done so my habits could be shared

**Comment by pflajszer** [comment]
+1 from me. I'd be very keen on either an exposed API, webhooks to subscribe to, or even some automated db exports.

**Comment by SimonDedman** [comment]
Hi, idk how useful this is but I recently noticed [Joplin](https://github.com/laurent22/joplin/) uses 3rd party services for backend sync Inc Dropbox, nextcloud, WebDAV, and many others. I wonder if you could copy their approach/code, they're open source?

**Comment by HelloLudger** [comment]
https://remotestorage.io/ could be used for this feature .

**Comment by AnsahMohammad** [comment]
Hey any update on this ?, 

About crossplatform, i was thinking how about writing Flutter code which can be used to build for other platforms (ios, linux, windows, web etc..), and it'll be a lot easier. Native for android can be maintained along with the flutter for rest.

About the database i think "supabase" is a good open source alternative

**Reply by jeff9315** [reply]
@AnsahMohammad ... Did you get anywhere with this? I've been toying with the idea of doing something like this with Flutter and [PowerSync](https://www.powersync.com/). Using either Postgress or MySQL ([Postgress support from PowerSync has just moved to Beta](https://www.powersync.com/blog/powersync-update-august-2025)).

**Comment by astingengo** [comment]
I would like to see this implemented to be able to have then web interface. Thank you for your amazing work

**Comment by kusk** [comment]
What about just changing the path in /uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/Constants.kt#L21 to something like /sdcard/Download or some other path that another sync app can handle? I changed the code ran it in a emulator and seems to work fine. However cant build a apk because of some config. Anyone else would be able to build it with those changes?

From:
`const val DATABASE_FILENAME = "uhabits.db"`

To:
`const val DATABASE_FILENAME = "/sdcard/Downloads/uhabits.db"`

**Reply by KyleSCraig** [reply]
Have a read of the note I made when building a few months ago and see if it unblocks you.

Did not get any links when trying to sink gradle, ultimately had to fix it by adding to the end of settings.gradle.kts:
plugins {
    id("org.gradle.toolchains.foojay-resolver-convention") version "0.4.0"
}

**Reply by EDLLT** [reply]
That's actually the simplest and best idea

**Reply by adharsh** [reply]
That's a great idea! I threw together a quick implementation here:
https://github.com/iSoron/uhabits/discussions/178#discussioncomment-11641659

It doesn't address point 3 in the original post tho, but it might be a step in the right direction

**Comment by EDLLT** [comment]
I found an open source tracker called ActivityWatch, their syncing mechanism is smart as it is service independent(leaving the user choose whatever method they find convenient whether that's gdrive, rsync, syncthing, ... through creating some sort of isolated directory structure.
 
I think uhabits could implement something similar
https://github.com/ActivityWatch/aw-server-rust/tree/master/aw-sync

**Comment by CHaP235** [comment]
I also +1 this idea/concept and a web based app would be much appreciated.

The only thing I can contribute to this discussion is to menton another Tracker I recently setup locally on my machine to further span the idea of having Loop self hosted/running locally:
https://github.com/redpangilinan/iotawise

It was a bit of a pain tbh but in the end; now it works! I'm not a pro but the Stack looks open-source (at least the database implementation), so maybe it's an inspiration for you guys to have another option.

**Comment by simone-viozzi** [comment]
Hi there, I came across this: https://github.com/daya0576/beaverhabits

Looks like the perfect backend for Loop, 
it already supports [importing from Loop](https://github.com/daya0576/beaverhabits/issues/1) and the web interface is similar to the Loop app.

**Comment by dashdeckers** [comment]
I think having a look at how Joplin handles this is the way to go. 

A first basic option should be available to synchronize data to a chosen (public) folder so that folks can use e.g. syncthing to manage synchronization themselves. 

This will already make a large number of users happy that use this app specifically because they value open source values and value owning their own data.

Then, incrementally add functionality to use OneDrive, DropBox, etc to manage synchronization / backups.

If no option is chosen, then the data lives as it currently does within a private folder within the android system.

**Reply by ghost** [reply]
Yes, I can manage my own sync, I just need to know what to sync!

**Comment by milan-tom** [comment]
@iSoron is there any plan regarding this? At present, even just adding an option to move the database to a public folder such that it can be synced with Syncthing would be a huge step forward. A more principled conflict resolution strategy could be introduced later, but this would significantly improve the experience right now with little effort. I can make a PR for this if you think this approach is valid.

**Comment by SimonDedman** [comment]
@milan-tom I'm not the boss of anything but in my experience in general, if you submit a PR, there's a decent chance it gets accepted - "come to me with solutions don't come to me with problems" principle. If you have the bandwidth, this would be great for everyone.

**Reply by milan-tom** [reply]
I think the approach is worth discussing before spending time on a PR. The [guidelines](https://github.com/iSoron/uhabits/blob/dev/docs/GUIDELINES.md) also suggest doing this for new features.

**Reply by KyleSCraig** [reply]
@milan-tom see #2209 I think it's already merged but not released.

**Reply by milan-tom** [reply]
@KyleSCraig if I am not mistaken, that PR is purely for automatically saving backup files to a public folder. Using that functionality for syncing would require an automatic import mechanism. The change I am suggesting is to store the raw internal files of the app in a public folder so they can be synced between devices. Then, no automatic import mechanism would be needed, since this is where the app reads from anyway.

**Comment by KyryloKaralyus** [comment]
@iSoron , wouldn't it be very simple to at least let users define where the db file is stored? That way, many power users could already start syncing (many cloud providers have sync/backup features !)
On the fancier side, have you seen the encrypted p2p sync protocol that "anytype"uses? Maybe that could work?

**Comment by otuva** [comment]
Hi,

I've been thinking about giving this a try as a hobby project, in my own time.

My idea is to build on the sync implementation that was removed in da018fc6, but replace the encrypted whole-database payload with a per-change event log, so offline edits from multiple devices can be merged properly, kinda how Joplin does it.

I'm guessing that limitation was why it was removed. The server would stay dumb, end-to-end encrypted, and self-hostable, so the three requirements from the original post should still hold. The merge logic would live in uhabits-core, so a future web or desktop client could reuse it.

Curious what people think of the approach and if you'd still use this, a reaction here helps me decide if it's worth the evenings :)
